apps/prairielearn/src/middlewares/workspaceProxy.ts

```diff
@@ -92,26 +131,34 @@ export function makeWorkspaceProxyMiddleware() {
 
       return `http://${result.rows[0].hostname}/`;
     },
-    onProxyReq: (proxyReq) => {
-      stripSensitiveCookies(proxyReq);
-    },
-    onProxyReqWs: (proxyReq) => {
-      stripSensitiveCookies(proxyReq);
-    },
-    onError: (err, req, res) => {
-      logger.error(`Error proxying workspace request: ${err}`, {
-        err,
-        url: req.url,
-        originalUrl: req.originalUrl,
-      });
-      // Check to make sure we weren't already in the middle of sending a
-      // response before replying with an error 500
-      if (res && !res.headersSent) {
-        res.status?.((err as any).status ?? 500)?.send?.('Error proxying workspace request');
-      }
+    on: {
+      proxyReq: (proxyReq) => {
+        stripSensitiveCookies(proxyReq);
+      },
+      proxyReqWs: (proxyReq) => {
+        stripSensitiveCookies(proxyReq);
+      },
+      error: (err, req, res) => {
+        logger.error(`Error proxying workspace request: ${err}`, {
+          err,
+          url: req.url,
+          originalUrl: req.originalUrl,
+        });
+
+        if (isResponseLike(res)) {
+          // Check to make sure we weren't already in the middle of sending a
+          // response before replying with our own error.
+          if (!res.headersSent) {
+            res.status(getStatusCode(err));
+          }
+
+          res.end('Error proxying workspace request');
+        } else if (isSocketLike(res)) {
+          // There's nothing we can do but destroy the socket.
+          res.destroy();
+        }
+      },
```

Comment:
The updated error handling here is based on my work from https://github.com/chimurai/http-proxy-middleware/pull/823 and is changed in anticipation of that. This error handler won't receive a `Socket` instance until we consume that change.